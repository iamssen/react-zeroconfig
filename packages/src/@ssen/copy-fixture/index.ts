import fs from 'fs-extra';
import path from 'path';

interface Params {
  cwd?: string;
  dest?: string;
}

export async function copyFixture(
  dir: string,
  {
    cwd = process.cwd(),
    dest = path.join(process.cwd(), '.fixtures'),
  }: Params = {},
): Promise<string> {
  const origin: string = path.join(cwd, dir);
  const fixture: string = path.join(dest, dir + '-' + Date.now());

  await fs.mkdirp(fixture);
  await fs.copy(origin, fixture);

  return fixture;
}
