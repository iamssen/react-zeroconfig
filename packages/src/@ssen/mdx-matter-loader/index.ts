import matter from 'gray-matter';

export default function (this: LoaderContext, contents: string) {
  const callback = this.async();

  if (!callback) return contents;

  try {
    const { content, data } = matter(contents);
    const decoded = JSON.stringify(data);
    const code = `${content}\n\nexport const frontmatter = ${decoded};`;

    callback(null, code);
  } catch {
    callback(null, contents);
  }
}

interface LoaderContext {
  async: () => (
    err: Error | null | undefined,
    content: string | Buffer | undefined,
  ) => void;
}
