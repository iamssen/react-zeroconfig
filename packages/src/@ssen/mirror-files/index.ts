import { Matcher } from 'anymatch';
import { watch } from 'chokidar';
import { FSWatcher } from 'fs';
import fs from 'fs-extra';
import path from 'path';
import { Observable } from 'rxjs';

interface Params {
  filesDirsOrGlobs: string[];
  outDir: string;
  ignored?: Matcher;
}

export type MirrorMessage =
  | {
      type: 'added' | 'updated' | 'removed';
      file: string;
      time: Date;
    }
  | {
      type: 'undefined';
      file: string;
      time: Date;
    };

export function mirrorFiles({
  filesDirsOrGlobs,
  outDir,
  ignored,
}: Params): Observable<MirrorMessage> {
  return new Observable<MirrorMessage>((subscriber) => {
    fs.mkdirpSync(outDir);

    function toRelativePath(file: string): string | undefined {
      const source: string | undefined = filesDirsOrGlobs.find(
        (s) => file.indexOf(s) === 0,
      );
      return source ? path.relative(source, file) : undefined;
    }

    const watcher: FSWatcher = watch(filesDirsOrGlobs, {
      ignored,
    });

    watcher
      //.on('ready', () => {
      //})
      .on('add', async (file) => {
        const relpath: string | undefined = toRelativePath(file);

        console.log('mirrorFiles index.ts..()', { relpath });

        if (!relpath) {
          subscriber.next({
            type: 'undefined',
            file,
            time: new Date(),
          });
          return;
        }

        const tofile: string = path.join(outDir, relpath);

        await fs.mkdirp(path.dirname(tofile));
        await fs.copy(file, tofile, { dereference: false });

        subscriber.next({
          type: 'added',
          file: relpath,
          time: new Date(),
        });
      })
      .on('change', async (file) => {
        const relpath: string | undefined = toRelativePath(file);

        if (!relpath) {
          subscriber.next({
            type: 'undefined',
            file,
            time: new Date(),
          });
          return;
        }

        const tofile: string = path.join(outDir, relpath);

        await fs.mkdirp(path.dirname(tofile));
        await fs.copy(file, tofile, { dereference: false });

        subscriber.next({
          type: 'updated',
          file: relpath,
          time: new Date(),
        });
      })
      .on('unlink', async (file) => {
        const relpath: string | undefined = toRelativePath(file);

        if (!relpath) {
          subscriber.next({
            type: 'undefined',
            file,
            time: new Date(),
          });
          return;
        }

        const tofile: string = path.join(outDir, relpath);

        if (fs.pathExistsSync(tofile)) {
          await fs.remove(tofile);

          subscriber.next({
            type: 'removed',
            file: relpath,
            time: new Date(),
          });
        }
      });

    return () => {
      watcher.close();
    };
  });
}
