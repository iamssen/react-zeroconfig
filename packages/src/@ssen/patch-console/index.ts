import { Console } from 'console';

/* eslint-disable @typescript-eslint/no-explicit-any */
type ConsoleMethodName = keyof Console;

const methodNames: ConsoleMethodName[] = Object.keys(console).filter(
  (method: string): method is keyof Console =>
    method !== 'Console' &&
    method !== 'context' &&
    typeof console[method as keyof Console] === 'function',
);
const originMethods: Map<ConsoleMethodName, any> = methodNames.reduce(
  (methods, name) => {
    methods.set(name, console[name]);
    return methods;
  },
  new Map(),
);

class ConsoleRouter implements Omit<Console, 'Console'> {
  private consoles: Set<Console> = new Set<Console>();

  add = (console: Console) => this.consoles.add(console);
  delete = (console: Console) => this.consoles.delete(console);
  size = () => this.consoles.size;

  memory = 0;
  exception = (...args: any[]) =>
    this.consoles.forEach((console) => console.exception(...args));
  assert = (...args: any[]) =>
    this.consoles.forEach((console) => console.assert(...args));
  clear = () => this.consoles.forEach((console) => console.clear());
  count = (...args: any[]) =>
    this.consoles.forEach((console) => console.count(...args));
  countReset = (...args: any[]) =>
    this.consoles.forEach((console) => console.countReset(...args));
  debug = (...args: any[]) =>
    this.consoles.forEach((console) => console.debug(...args));
  dir = (...args: any[]) =>
    this.consoles.forEach((console) => console.dir(...args));
  dirxml = (...args: any[]) =>
    this.consoles.forEach((console) => console.dirxml(...args));
  error = (...args: any[]) =>
    this.consoles.forEach((console) => console.error(...args));
  group = (...args: any[]) =>
    this.consoles.forEach((console) => console.group(...args));
  groupCollapsed = (...args: any[]) =>
    this.consoles.forEach((console) => console.groupCollapsed(...args));
  groupEnd = () => this.consoles.forEach((console) => console.groupEnd());
  info = (...args: any[]) =>
    this.consoles.forEach((console) => console.info(...args));
  log = (...args: any[]) =>
    this.consoles.forEach((console) => console.log(...args));
  table = (...args: any[]) =>
    this.consoles.forEach((console) => console.table(...args));
  time = (...args: any[]) =>
    this.consoles.forEach((console) => console.time(...args));
  timeEnd = (...args: any[]) =>
    this.consoles.forEach((console) => console.timeEnd(...args));
  timeLog = (...args: any[]) =>
    this.consoles.forEach((console) => console.timeLog(...args));
  trace = (...args: any[]) =>
    this.consoles.forEach((console) => console.trace(...args));
  warn = (...args: any[]) =>
    this.consoles.forEach((console) => console.warn(...args));
  profile = (...args: any[]) =>
    this.consoles.forEach((console) => console.profile(...args));
  profileEnd = (...args: any[]) =>
    this.consoles.forEach((console) => console.profileEnd(...args));
  timeStamp = (...args: any[]) =>
    this.consoles.forEach((console) => console.timeStamp(...args));
}

let router: ConsoleRouter | null;

export function patchConsole(
  options: NodeJS.ConsoleConstructorOptions,
): () => void {
  const childConsole = new Console(options);

  if (!router) {
    router = new ConsoleRouter();

    for (const method of methodNames) {
      console[method] = router[method as keyof ConsoleRouter];
    }
  }

  router.add(childConsole);

  return () => {
    if (!router) return;

    router.delete(childConsole);

    if (router.size() === 0) {
      for (const method of methodNames) {
        console[method] = originMethods.get(method);
      }

      router = null;
    }
  };
}
