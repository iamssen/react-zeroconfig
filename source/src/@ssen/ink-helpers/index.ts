import { EventEmitter } from 'events';
import { render } from 'ink';
import { ReactElement } from 'react';

export class InkWritableStream extends EventEmitter {
  get columns() {
    return 100;
  }

  readonly frames: string[] = [];
  private _lastFrame?: string;

  write = (frame: string) => {
    this.frames.push(frame);
    this._lastFrame = frame;
  };

  lastFrame = () => {
    return this._lastFrame;
  };
}

export function createInkWriteStream(): InkWritableStream & NodeJS.WriteStream {
  return (new InkWritableStream() as unknown) as InkWritableStream & NodeJS.WriteStream;
}

export function inkToString(element: ReactElement): string {
  const stdout = createInkWriteStream();

  const { unmount } = render(element, { stdout, patchConsole: false });

  unmount();

  return stdout.lastFrame() ?? '';
}
