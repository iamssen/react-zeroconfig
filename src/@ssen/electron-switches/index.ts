import type { Argv } from 'yargs';

type Options = Parameters<Argv['options']>[0];

const describe = '* electon command-line switch';

/**
 * @see https://www.electronjs.org/docs/api/command-line-switches
 */
export const electronSwitchesYargsOptions = {
  'ignore-connections-limit': { type: 'string', describe },
  'disable-http-cache': { type: 'boolean', describe },
  'disable-http2': { type: 'boolean', describe },
  'disable-ntlm-v2': { type: 'boolean', describe },
  lang: { type: 'string', describe },
  inspect: { type: 'string', describe },
  'inspect-brk': { type: 'string', describe },
  'remote-debugging-port': { type: 'number', describe },
  'disk-cache-size': { type: 'number', describe },
  'js-flags': { type: 'string', describe },
  'proxy-server': { type: 'string', describe },
  'proxy-bypass-list': { type: 'string', describe },
  'proxy-pac-url': { type: 'string', describe },
  'no-proxy-server': { type: 'boolean', describe },
  'host-rules': { type: 'string', describe },
  'host-resolver-rules': { type: 'string', describe },
  'auth-server-whitelist': { type: 'string', describe },
  'auth-negotiate-delegate-whitelist': { type: 'string', describe },
  'ignore-certificate-errors': { type: 'boolean', describe },
  'log-net-log': { type: 'string', describe },
  'disable-renderer-backgrounding': { type: 'boolean', describe },
  'enable-logging': { type: 'boolean', describe },
  v: { type: 'string', describe },
  vmodule: { type: 'string', describe },
  'enable-api-filtering-logging': { type: 'boolean', describe },
  'no-sandbox': { type: 'boolean', describe },
} as const;

export type ElectronSwitchesYargsValues = {
  [Key in keyof typeof electronSwitchesYargsOptions]?: typeof electronSwitchesYargsOptions[Key]['type'] extends 'boolean'
    ? boolean
    : typeof electronSwitchesYargsOptions[Key]['type'] extends 'number'
    ? number
    : string;
};

export function toElectronArgv(values: ElectronSwitchesYargsValues): string[] {
  return Object.keys(values).reduce((args: string[], key) => {
    if (!!electronSwitchesYargsOptions[key]) {
      switch (electronSwitchesYargsOptions[key].type) {
        case 'boolean':
          if (values[key] === true) {
            args.push(`--${key}`);
          }
          break;
        case 'number':
          args.push(`--${key}=${values[key]}`);
          break;
        default:
          args.push(`--${key}="${values[key]}"`);
          break;
      }
    }
    return args;
  }, []);
}
