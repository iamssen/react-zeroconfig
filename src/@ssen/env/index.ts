import path from 'path';
import fs from 'fs';

export const mapEnv: (envFile: string) => (env: NodeJS.ProcessEnv) => NodeJS.ProcessEnv = (envFile) => (env) => {
  const file: string = path.join(process.cwd(), envFile);

  if (!fs.existsSync(file)) return env;

  return require(file)(env);
};

export const patchEnv: (origin?: NodeJS.ProcessEnv) => (env: NodeJS.ProcessEnv) => void = (origin = {}) => (env) => {
  Object.keys(env).forEach((key) => {
    if (env[key] && env[key] !== origin[key]) {
      process.env[key] = env[key];
    }
  });
};

export const printEnv: (...keys: string[]) => (env: NodeJS.ProcessEnv) => void = (...keys) => (env) => {
  const picked: NodeJS.ProcessEnv = Object.keys(env)
    .filter((key) => keys.indexOf(key) > -1)
    .reduce((picked, key) => {
      picked[key] = env[key];
      return picked;
    }, {});

  console.log(JSON.stringify(picked, null, 2));
};
