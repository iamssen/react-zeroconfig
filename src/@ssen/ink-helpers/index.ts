import { EventEmitter } from 'events';
import { render } from 'ink';
import { ReactElement } from 'react';

export class InkWritableStream extends EventEmitter {
  get columns() {
    return 100;
  }

  readonly frames: string[] = [];
  private _lastFrame?: string;

  write = (frame: string) => {
    this.frames.push(frame);
    this._lastFrame = frame;
  };

  lastFrame = () => {
    return this._lastFrame;
  };
}

export function createInkWriteStream(): InkWritableStream & NodeJS.WriteStream {
  return (new InkWritableStream() as unknown) as InkWritableStream & NodeJS.WriteStream;
}

export function inkToString(element: ReactElement): string {
  const stdout = createInkWriteStream();

  const { unmount } = render(element, { stdout });

  unmount();

  return stdout.lastFrame() ?? '';
}
