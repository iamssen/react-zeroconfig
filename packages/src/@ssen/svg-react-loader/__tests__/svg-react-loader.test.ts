import { copyFixture } from '@ssen/copy-fixture';
import svgReactLoader from '@ssen/svg-react-loader';
import path from 'path';

function compile(
  resourcePath: string,
  source: string,
): Promise<string | Buffer | undefined> {
  return new Promise<string | Buffer | undefined>((resolve, reject) => {
    const context = {
      resourcePath,
      async: () => (
        err: Error | null | undefined,
        content: string | Buffer | undefined,
      ) => {
        if (err) {
          reject(err);
        } else {
          resolve(content);
        }
      },
    };

    svgReactLoader.call(context, source);
  });
}

describe('svg-react-loader', () => {
  test('should get react jsx codes', async () => {
    // Arrange
    const cwd: string = await copyFixture(`test/fixtures/web/bundle`);

    const content = ['export default "/absoulte-path/test.svg"'].join('\n');

    const result = await compile(
      path.join(cwd, 'src/app/components/test.svg'),
      content,
    );

    expect(result).toMatchSnapshot();
  });
});
