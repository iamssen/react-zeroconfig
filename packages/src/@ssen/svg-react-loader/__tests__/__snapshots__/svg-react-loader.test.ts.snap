// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svg-react-loader should get react jsx codes 1`] = `
"import * as React from \\"react\\";

function ReactComponent() {
  return <svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\"><defs><style>{\\".a{fill:#5269e8;}.b{fill:#ad450e;}.c{fill:none;}\\"}</style></defs><g transform=\\"translate(-1877 -1486)\\"><path className=\\"a\\" d=\\"M12,24A12,12,0,0,1,3.515,3.515,12,12,0,1,1,12,24ZM12,3.428A8.572,8.572,0,1,0,20.572,12,8.582,8.582,0,0,0,12,3.428Z\\" transform=\\"translate(1877 1486)\\" /><path className=\\"b\\" d=\\"M6,12a6,6,0,1,1,6-6A6.007,6.007,0,0,1,6,12ZM6,3.272A2.728,2.728,0,1,0,8.728,6,2.731,2.731,0,0,0,6,3.272Z\\" transform=\\"translate(1883 1492)\\" /></g></svg>;
}

export {ReactComponent};
export default \\"/absoulte-path/test.svg\\""
`;
